src/spec.rs

```diff
@@ -0,0 +1,109 @@
+//! Helper functions defined in the Zcash Protocol Specification.
+
+use std::iter;
+
+use blake2b_simd::Params;
+use ff::PrimeField;
+use group::{Curve, Group};
+use halo2::{
+    arithmetic::{CurveAffine, CurveExt, FieldExt},
+    pasta::pallas,
+};
+
+use crate::{constants::L_ORCHARD_BASE, primitives::sinsemilla};
+
+const PRF_EXPAND_PERSONALIZATION: &[u8; 16] = b"Zcash_ExpandSeed";
+
+/// $\mathsf{ToBase}^\mathsf{Orchard}(x) := LEOS2IP_{\ell_\mathsf{PRFexpand}}(x) (mod q_P)$
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+pub(crate) fn to_base(x: [u8; 64]) -> pallas::Base {
+    pallas::Base::from_bytes_wide(&x)
+}
+
+/// $\mathsf{ToScalar}^\mathsf{Orchard}(x) := LEOS2IP_{\ell_\mathsf{PRFexpand}}(x) (mod r_P)$
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+pub(crate) fn to_scalar(x: [u8; 64]) -> pallas::Scalar {
+    pallas::Scalar::from_bytes_wide(&x)
+}
+
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+pub(crate) fn commit_ivk(
+    ak: &pallas::Base,
+    nk: &pallas::Base,
+    rivk: &pallas::Scalar,
+) -> pallas::Scalar {
+    // We rely on the API contract that to_le_bits() returns at least PrimeField::NUM_BITS
+    // bits, which is equal to L_ORCHARD_BASE.
+    let ivk = sinsemilla::short_commit(
+        "z.cash:Orchard-CommitIvk",
+        iter::empty()
+            .chain(ak.to_le_bits().iter().by_val().take(L_ORCHARD_BASE))
+            .chain(nk.to_le_bits().iter().by_val().take(L_ORCHARD_BASE)),
+        rivk,
+    );
+
+    // Convert from pallas::Base to pallas::Scalar. This requires no modular reduction
+    // because Pallas' base field is smaller than its scalar field.
+    pallas::Scalar::from_repr(ivk.to_repr()).unwrap()
+}
+
+/// Defined in [Zcash Protocol Spec § 5.4.1.6: DiversifyHash^Sapling and DiversifyHash^Orchard Hash Functions][concretediversifyhash].
+///
+/// [concretediversifyhash]: https://zips.z.cash/protocol/nu5.pdf#concretediversifyhash
+pub(crate) fn diversify_hash(d: &[u8; 11]) -> Option<pallas::Point> {
+    let pk_d = pallas::Point::hash_to_curve("z.cash:Orchard-gd")(d);
+    if pk_d.is_identity().into() {
+        None
```

Comment:
In my pairing with @ebfull, we discussed this, and it occurred to me that it's possible we actually never hit this: once we fix Pallas and the hash-to-curve domain separator, the diversifier space is fixed for all spending keys. Thus any diversifier that resulted in the identity would do so for _all_ spending keys (though the spending key's diversifier key would alter which index that occured at, which could affect the default address).

We could in theory brute-force the 88-bit space to confirm that no diversifier will result in the identity. It's not practical to do so, but given that it's an 88-bit value being hashed into a 255-bit space, I think it's probably reasonable to just assert we never hit it. Thoughts @daira?

Reply:
Another thing we could do is that if the result of the group hash is the zero point, we just replace it with another arbitrary point, such as another group hash output or (-1, 2). I would mildly prefer that.

Reply:
![image](https://user-images.githubusercontent.com/643204/111509640-d66e2280-8744-11eb-8cb5-5200f29d75d0.png)

Reply:
The type of DiversifyHash should now be B -> P* (without the bottom).

Reply:
Caught it in 4.1.1 but not in 5.4.1.6. That'll be in the next spec revision.